From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: marcus hall <marcus.hall@caci.com>
Date: Fri, 12 Jan 2024 10:05:58 -0700
Subject: [PATCH] 6113: Fix branch sense in "bany" instruction. Add special
 cases for "sext" of bytes.

clamps instruction should put result in ar, not just a temporary.
---
 .../Xtensa/data/languages/xtensaInstructions.sinc   | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/Ghidra/Processors/Xtensa/data/languages/xtensaInstructions.sinc b/Ghidra/Processors/Xtensa/data/languages/xtensaInstructions.sinc
index 1aa19cb0d..9e7ed1e2b 100644
--- a/Ghidra/Processors/Xtensa/data/languages/xtensaInstructions.sinc
+++ b/Ghidra/Processors/Xtensa/data/languages/xtensaInstructions.sinc
@@ -122,7 +122,7 @@
 # BANY - Branch if Any Bit Set, pg. 265.
 :bany srel_16_23, as, at, is srel_16_23 & ar = 0b1000 & as & at & op0 = 0b0111 {
 	local test:4 = as & at;
-	if (test == 0)	goto srel_16_23;
+	if (test != 0)	goto srel_16_23;
 }
 
 macro extract_bit(bit, result) {
@@ -399,7 +399,7 @@ macro extract_bit(bit, result) {
 	local mt:1 = (x s> (-clamp));
 	local max:4 = (zext(mt) * x) + (zext(!mt) * (-clamp));
 	mt = (x s< (clamp-1));
-	local y = (zext(mt) * max) + (zext(!mt) * (clamp-1));
+	ar = (zext(mt) * max) + (zext(!mt) * (clamp-1));
 }
 
 # DHI - Data Cache Hit Invalidate, pg. 313.
@@ -1166,6 +1166,15 @@ rfi_eps: ptr is u4_8_11	[ ptr = $(EPS_BASE) + (4 * u4_8_11); ] { export *[regist
 	local tmp:4 = as << shift;
 	ar = tmp s>> shift;
 }
+:sext ar, as, 7 is op2 = 0b0010 & op1 = 0b0011 & ar & as & u4_4_7 = 0 & op0 = 0 {
+    ar = sext(as:1);
+}
+:sext ar, as, 15 is op2 = 0b0010 & op1 = 0b0011 & ar & as & u4_4_7 = 8 & op0 = 0 {
+    ar = sext(as:2);
+}
+:sext ar, as, 23 is op2 = 0b0010 & op1 = 0b0011 & ar & as & u4_4_7 = 12 & op0 = 0 {
+    ar = sext(as:3);
+}
 
 # SICT - Store Instruction Cache Tag, pg. 519.
 :sict at, as is op2 = 0b1111 & op1 = 0b0001 & ar = 0b0001 & as & at & op0 = 0 {
-- 
2.43.0

